### PR TITLE
Add Value JProfiling

### PR DESCRIPTION
**Pull request by ncough**
This adds a new value profiling implementation,
capable of collecting information with low
overhead. It achieves this by only tracking
the most frequent values and stores them in
a hash map to simplify lookups. A majority
of the instrumentation is represented in
trees.

This is used in an optimization pass, capable
of identifying common profiling candidates in
arbitrary compilations, as well as in a modified
ValueProfiler, used to instrument profiling
compilations.

### CODE REVIEW

File: runtime/compiler/trj9/control/J9Recompilation.cpp [hunk 1]
```diff
@@ -439,6 +442,24 @@ J9::Recompilation::getValueProfiler()
...
+J9::Recompilation::getBlockFrequencyProfiler()
+   {
+   for (TR_RecompilationProfiler * rp = getFirstProfiler(); rp; rp = rp->getNext())
+      {
+      TR_BlockFrequencyProfiler *vp = rp->asBlockFrequencyProfiler();
+      if (vp)
+         return vp;
+      }
+
+   return 0;
```
Comment (andrewcraik) [inline runtime/compiler/trj9/control/J9Recompilation.cpp]: This should use return NULL rather than return 0 for future compatibility.

File: runtime/compiler/trj9/control/J9Recompilation.cpp [hunk 2]
```diff
@@ -439,6 +442,24 @@ J9::Recompilation::getValueProfiler()
    return 0;
    }
 
+TR_BlockFrequencyProfiler *
+J9::Recompilation::getBlockFrequencyProfiler()
+   {
+   for (TR_RecompilationProfiler * rp = getFirstProfiler(); rp; rp = rp->getNext())
+      {
+      TR_BlockFrequencyProfiler *vp = rp->asBlockFrequencyProfiler();
+      if (vp)
```
Comment (andrewcraik) [inline runtime/compiler/trj9/control/J9Recompilation.cpp]: A comment about why this can be NULL is probably warranted

File: runtime/compiler/trj9/control/J9Recompilation.cpp [hunk 3]
```diff
@@ -439,6 +442,24 @@ J9::Recompilation::getValueProfiler()
    return 0;
    }
 
+TR_BlockFrequencyProfiler *
+J9::Recompilation::getBlockFrequencyProfiler()
```
Comment (andrewcraik) [inline runtime/compiler/trj9/control/J9Recompilation.cpp]: As a new API this should have doxygen

File: runtime/compiler/trj9/control/J9Recompilation.cpp [hunk 4]
```diff
@@ -439,6 +442,24 @@ J9::Recompilation::getValueProfiler()
...
+      TR_BlockFrequencyProfiler *vp = rp->asBlockFrequencyProfiler();
+      if (vp)
+         return vp;
+      }
+
+   return 0;
+   }
+
+void
+J9::Recompilation::removeProfiler(TR_RecompilationProfiler *rp)
```
Comment (andrewcraik) [inline runtime/compiler/trj9/control/J9Recompilation.cpp]: doxygen for this as well

File: runtime/compiler/trj9/runtime/J9Profiler.hpp [hunk 5]
```diff
@@ -205,6 +206,7 @@ class TR_RecompilationProfiler : public TR_Link<TR_RecompilationProfiler>
    void setInitialiCompilation(bool b) { _flags.set(initialCompilation, b); }
 
    virtual TR_ValueProfiler * asValueProfiler() {return NULL;}
+   virtual TR_BlockFrequencyProfiler * asBlockFrequencyProfiler() {return NULL;}
```
Comment (andrewcraik) [inline runtime/compiler/trj9/runtime/J9Profiler.hpp]: code style - prefer spaces around { }s

File: runtime/compiler/trj9/optimizer/BlockJProfiling.cpp [hunk 6]
```diff
@@ -756,42 +756,18 @@ int32_t TR_BlockJProfiling::processCFGForCounting(BlockParents &parent, TR::Bloc
...
-   else if (profileInfo->getCallSiteInfo()->getNumCallSites() != comp()->getNumInlinedCallSites())
-      {
-      TR_CallSiteInfo * const originalCallSiteInfo = profileInfo->getCallSiteInfo();
-      TR_ASSERT(originalCallSiteInfo != NULL, "Existing CallSiteInfo should not be NULL.");
-      TR_CallSiteInfo * const updatedCallSiteInfo = new (PERSISTENT_NEW) TR_CallSiteInfo(comp(), persistentAlloc);
-      profileInfo->setCallSiteInfo(updatedCallSiteInfo);
-      // FIXME: originalCallSiteInfo and its _blocks array allocation appear to leak.
-      }
 
+   // Remove other block frequency profiler
```
Comment (andrewcraik) [inline runtime/compiler/trj9/optimizer/BlockJProfiling.cpp]: what other block frequency profiler? this seems a bit cryptic

File: runtime/compiler/trj9/runtime/JitRuntime.cpp [hunk 7]
```diff
@@ -1155,6 +1169,8 @@ JIT_HELPER(_jitProfileLongValueX);
 JIT_HELPER(_jitProfileBigDecimalValueX);
 JIT_HELPER(_jitProfileStringValueX);
 JIT_HELPER(_jitProfileAddressX);
+JIT_HELPER(_jProfileValueX);
+JIT_HELPER(_jProfileLongValueX);
```
Comment (fjeremic) [inline runtime/compiler/trj9/runtime/JitRuntime.cpp]: Would it be more clear if these were `_jProfile32BitValue` and `_jProfileValue64BitValue` instead? The names seem orthogonal at the moment.

Comment (ncough) [reply runtime/compiler/trj9/runtime/JitRuntime.cpp]: Thanks for the suggestion, I have updated them accordingly.

**Comment by andrewcraik** [comment]
@ncough genuine failures in the travis build - looks like some kind of C++ issue (perhaps a GCC version difference is a cause or some other quirk).

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 8]
```diff
@@ -0,0 +1,883 @@
...
+      }
+
+   TR_ValueProfileInfo *valueProfileInfo = TR_PersistentProfileInfo::get(comp())->findOrCreateValueProfileInfo(comp());
+   TR_AbstractHashTableProfilerInfo *info = (TR_AbstractHashTableProfilerInfo*) valueProfileInfo->getOrCreateProfilerInfo(address->getByteCodeInfo(), comp(), AddressInfo, HashTableProfiler);
+
+   addProfilingTrees(comp(), tt, vftNode, info, check, fallback, true, trace());
+   }
+
+/**
+ * Identify helper calls to TR_jProfileValue and TR_jProfileLongValue, lowering them
```
Comment (fjeremic) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: IIRC these names were changed. I think this comment should be updated.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 9]
```diff
@@ -0,0 +1,883 @@
...
+ * | ...               |
+ *
+ * \param insertionPoint Treetop to insert profiling code after.
+ * \param value Value to profile.
+ * \param table Persistent TR_HashMapInfo which will be filled and incremented during profiling.
+ * \param optionalTest Option test node capable of preventing evaluation of value and using a fallbackValue instead.
+ * \param fallbackValue Fallback value to use with the optional test.
+ * \param extendBlocks Generates the blocks as extended, defaults true.
+ * \param trace Enable tracing.
+ */
```
Comment (fjeremic) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: Awesome documentation! Readers take note 😄

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 10]
```diff
@@ -0,0 +1,883 @@
...
+      traceMsg(comp, "Inserting value profiling trees:\n  Value n%dn At n%dn\n  Table: %p\n",
+         value->getGlobalIndex(),
+         insertionPoint->getNode()->getGlobalIndex(),
+         table);
+      table->trace(comp);
+      traceMsg(comp, "  Test: %p Fallback: %p\n", optionalTest, fallbackValue);
+      }
+
+   TR::Block    *originalBlock = insertionPoint->getEnclosingBlock();
+   TR::Block    *extendedBlock = originalBlock;
```
Comment (fjeremic) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: Unnecessary spaces here.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 11]
```diff
@@ -0,0 +1,883 @@
...
+   while (tt && tt != end)
+      {
+      TR::Node *node = tt->getNode();
+      if (node->getOpCodeValue() != TR::BBStart && node->getOpCodeValue() != TR::BBEnd)
+         setProfilingCode(node, checklist);
+      tt = tt->getNextTreeTop();
+      }
+   tt = slowTest->getEntry();
+   end = NULL;
+   while (tt && tt != end)
```
Comment (fjeremic) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: Should be able to simplify this to `while (tt)` and remove line above.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 12]
```diff
@@ -0,0 +1,883 @@
...
+ * Given an address in X86's LEA form, generate a node structure to calculate a memory address for a desired
+ * data type. Address would be: base + width(dataType) * index + offset
+ *
+ * \param dataType Date type of value to address. Its width is used for calculations if index is specified.
+ * \param base Base address to load from.
+ * \param index Optional index.
+ * \param offset Optional offset from the base.
+ */
+TR::Node *
+TR_ValueJProfiling::effectiveAddress(TR::DataType dataType, TR::Node *base, TR::Node *index, TR::Node *offset)
```
Comment (fjeremic) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: Seems oddly x86 specific. Will this code run for other platforms as well?

File: runtime/compiler/trj9/optimizer/BlockJProfiling.cpp [hunk 13]
```diff
@@ -971,8 +958,42 @@ void TR_JProfiling::addRecompilationTests(TR_BlockFrequencyInfo *blockFrequencyI
...
+    * BlockJProfiling will run early in the compilation, such that the CFG closely resembles the original program structure,
+    * whilst the ValueJProfiling pass runs later, as values may be introduced, eliminated or simplified.
+    *
+    * For non-profiling compilations, this can be enabled with the option TR_EnableJProfiling. Other control infrastructure
+    * may limit which compilations this is actually applied to. See the env option TR_DisableFilterOnJProfiling for more details.
+    *
+    * For profiling compilations, JProfiling can replace the existing JitProfiling instrumentation when either TR_EnableJProfiling
+    * or TR_EnableJProfilingInProfilingCompilations are set. However, this introduces some complexity as a compilation may
+    * switch to profiling part way though, potentially after BlockJProfiling should have run. In such a scenario, profiling compilations
+    * will ignore JProfiling and fallback on the prior implementation.
```
[the remark this answers is not in the document]
Comment (andrewcraik) [reply runtime/compiler/trj9/optimizer/BlockJProfiling.cpp]: I believe the proper fix in this case is to abort the compilation and restart as a profiling compilation. The mid-compilation switch was always non-ideal and JProfiling just highlights why this is the case. I don't view this as a blocker. I believe @ncough had some conversations with @mpirvu to the effect that we all agree an abort and restart is warranted and that will avoid the whole mess. The comment can go once that is implemented.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 14]
```diff
@@ -0,0 +1,883 @@
...
+#include "il/Node_inlines.hpp"
+#include "infra/Checklist.hpp"             // for TR::NodeChecklist
+#include "ras/DebugCounter.hpp"
+#include "runtime/J9Profiler.hpp"
+#include "control/Recompilation.hpp"              // for TR_Recompilation, etc
+#include "control/RecompilationInfo.hpp"              // for TR_Recompilation, etc
+#include "codegen/CodeGenerator.hpp"
+#include "optimizer/TransformUtil.hpp"            // for TransformUtil
+
+#define COLD_BLOCK_FREQ 6
```
[the remark this answers is not in the document]
Comment (andrewcraik) [reply runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: That frequency should already be defined else where as a macro so we should use that global definition rather than duplicating the value of 6 either as a static const or a macro.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 15]
```diff
@@ -0,0 +1,883 @@
...
+
+// There are existing versions of these mappings, however, they will round up
+// types smaller than Int32
+TR::ILOpCodes
+directStore(TR::DataType dt)
```
[the remark this answers is not in the document]
Comment (andrewcraik) [reply runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: The key here is the note about the round up - in Java sub-int sized values will be widened to int - this needs the specific width. This could potentially go somewhere else, but I am fine with it being here since it is an anomaly. I don't think we should be reaching up to OMR directly rather than going via the TR mapping - it is a dangerous precedent and leaves this code open to being broken by changes in OMR that could be completely unreleated. Java chose to replace the OMR definitions so this code needs to handle things its own way IMO. When we go to contribute this up to OMR then we can revisit the design.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp (hunk 12 above)
Comment (andrewcraik) [reply runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: I will let @ncough comment, but I don't think this style is a problem on other platforms - it exposes maximum richness and the optimizer will have an opportunity to rearrange if necessary as will the code generators.

File: runtime/compiler/trj9/optimizer/BlockJProfiling.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (mpirvu) [reply runtime/compiler/trj9/optimizer/BlockJProfiling.cpp]: As long as the number of such aborts/restarts is low we should be fine.
If SCC is enabled we should implement profiling hints such that future JVM invocations will be able to trigger the profiling compilation from the beginning.

[the remark this answers is not in the document]
Comment (andrewcraik) [reply runtime/compiler/trj9/optimizer/BlockJProfiling.cpp]: Ideally, we would be able to put the profiling code into AOT code and relocate the counters so we don't even need that, but I agree one step at a time. Excessive aborts would be problematic, but given the low use of profiling compilations atm we should be fine. We should make sure that the vlog calls out such aborts very clearly so these events are easy to track and diagnose if a problem does occur.

File: runtime/compiler/trj9/codegen/J9TreeEvaluator.cpp [hunk 16]
```diff
@@ -48,7 +48,7 @@ void traceInstanceOfOrCheckCastProfilingInfo(TR::CodeGenerator *cg, TR::Node *no
       return;
       }
 
-   TR_AddressInfo * valueInfo = (TR_AddressInfo *)valueProfileInfo->getValueInfo(bcInfo, comp, TR_ValueProfileInfoManager::justInterpreterProfileInfo);
+   TR_AddressInfo * valueInfo = (TR_AddressInfo *)valueProfileInfo->getValueInfo(bcInfo, comp, AddressInfo, TR_ValueProfileInfoManager::justInterpreterProfileInfo);
```
Comment (andrewcraik) [inline runtime/compiler/trj9/codegen/J9TreeEvaluator.cpp]: A static cast would be better here no?

File: runtime/compiler/trj9/compile/J9Compilation.hpp [hunk 17]
```diff
@@ -164,8 +164,7 @@ class OMR_EXTENSIBLE Compilation : public OMR::CompilationConnector
 
    bool pendingPushLivenessDuringIlgen();
 
-   TR::list<TR_MethodValueProfileInfo*> &getMethodVPInfos() { return _methodVPInfoList; }
-   TR::list<TR_MethodValueProfileInfo*> &getMethodHWVPInfos() { return _methodHWVPInfoList; }
+   TR::list<TR_ExternalValueProfileInfo*> &getExternalVPInfos() { return _externalVPInfoList; }
```
Comment (andrewcraik) [inline runtime/compiler/trj9/compile/J9Compilation.hpp]: A discussion comment in reply to this comment on why HWVPInfos can go - I assume it came from hardware profiling work and I'd just like to understand how we can clean this up without breaking anything.

File: runtime/compiler/trj9/runtime/IProfiler.cpp [hunk 18]
```diff
@@ -1548,7 +1548,7 @@ TR_IProfiler::profilingSample (TR_OpaqueMethodBlock *method, uint32_t byteCodeIn
       if (!preferHashtableData)
          {
          // If I don't have data in IProfiler HT, choose the persistent source
-         if(!currentEntry || (currentEntry->getData() == NULL))
+         if(!currentEntry || (currentEntry->getData() == (uintptr_t) NULL))
```
Comment (andrewcraik) [inline runtime/compiler/trj9/runtime/IProfiler.cpp]: no space before NULL

File: runtime/compiler/trj9/runtime/IProfiler.cpp [hunk 19]
```diff
@@ -1561,7 +1561,7 @@ TR_IProfiler::profilingSample (TR_OpaqueMethodBlock *method, uint32_t byteCodeIn
                }
             }
          // If I don't have relevant data in the SCC, choose the data from IProfiler HT
-         else if(!persistentEntry || (persistentEntry->getData() == NULL))
+         else if(!persistentEntry || (persistentEntry->getData() == (uintptr_t) NULL))
```
Comment (andrewcraik) [inline runtime/compiler/trj9/runtime/IProfiler.cpp]: no space before NULL

File: runtime/compiler/trj9/runtime/J9Profiler.cpp [hunk 20]
```diff
@@ -894,29 +881,50 @@ TR_ValueProfileInfoManager::TR_ValueProfileInfoManager(TR::Compilation *comp)
...
+TR_ValueProfileInfoManager::getValueInfo(TR_ByteCodeInfo &bcInfo, TR::Compilation *comp, TR_ValueInfoKind kind,
+      uint32_t source)
    {
    TR_AbstractInfo *info = NULL;
+   bool internal = _jitValueProfileInfo && (source == allProfileInfo || source == justJITProfileInfo);
+   bool external = source == allProfileInfo || source == justInterpreterProfileInfo;
 
-   if (_jitValueProfileInfo && (profileInfoKind != justInterpreterProfileInfo))
-      info = _jitValueProfileInfo->getValueInfo(bcInfo, comp, type);
+   if ((!info || info->getTotalFrequency() == 0) && internal)
```
Comment (andrewcraik) [inline runtime/compiler/trj9/runtime/J9Profiler.cpp]: in general it would be more neat to do internal && (....). That would have the benefit of testing the already computed bool first before potentially doing a deref which could be a savings if this runs a lot.

File: runtime/compiler/trj9/runtime/J9ValueProfiler.hpp [hunk 21]
```diff
@@ -28,266 +28,842 @@
...
+ *       TR_AddressInfo
+ *       TR_BigDecimalInfo
+ */
+
+/**
+ * Kinds of value profiling information
+ */
+enum TR_ValueInfoKind
+   {
+   ValueInfo            = 0x00,
```
Comment (andrewcraik) [inline runtime/compiler/trj9/runtime/J9ValueProfiler.hpp]: If these are not being used as individual bits why do we need to assign numbers in a flags style? Can't we use the default numbering or standard decimal numbering so we don't think a bit or will do something useful which is the usual intent between a hex value assignment eg 0x1, 0x2, 0x4, 0x8 etc.

File: runtime/compiler/trj9/runtime/J9ValueProfiler.hpp [hunk 22]
```diff
@@ -28,266 +28,842 @@
...
-class TR_ExtraAbstractInfo
+/**
+ * Sources of JIT value profiling information
+ * This excludes external sources, such as HW and the interpreter
+ */
+enum TR_ValueInfoSource
    {
-   public:
-   TR_ALLOC(TR_Memory::PersistentProfileInfo)
+   LinkedListProfiler = 0x00,
```
Comment (andrewcraik) [inline runtime/compiler/trj9/runtime/J9ValueProfiler.hpp]: Same as last enum

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (andrewcraik) [reply runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: Doxygen on these methods and the reason for their split from OMR is necessary.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 23]
```diff
@@ -0,0 +1,883 @@
...
+   {
+   TR::SymbolReference *profiler;
+   if (value->getDataType() == TR::Int32)
+      profiler = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jProfile32BitValue, false, false, true);
+   else
+      profiler = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jProfile64BitValue, false, false, true);
+
+#if defined(TR_HOST_POWER) || defined(TR_HOST_ARM)
+   profiler->getSymbol()->castToMethodSymbol()->setLinkage(TR_System);
+#else
```
Comment (andrewcraik) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: this seems ugly can we avoid the nested case here and make this a more clean if elif else?

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 24]
```diff
@@ -0,0 +1,883 @@
...
+/*
+ * Its sometimes necessary to convert values to their rounded integer representation, in Int32
+ * or Int64, or to convert values to the system type, again in Int32 or Int64. This method
+ * handles those conversions as necessary. All conversions are unsigned.
+ *
+ * \param index The value to convert.
+ * \param dataType The datatype to convert to.
+ */
+TR::Node *
+TR_ValueJProfiling::convertType(TR::Node *index, TR::DataType dataType)
```
Comment (andrewcraik) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: There isn't an OMR helper for this that could be used?

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp [hunk 25]
```diff
@@ -0,0 +1,883 @@
...
+   return NULL;
+   }
+
+TR::Node *
+TR_ValueJProfiling::computeHash(TR::Compilation *comp, TR_AbstractHashTableProfilerInfo *table, TR::Node *value, TR::Node *baseAddr)
+   {
+   if (!baseAddr)
+      baseAddr = TR::Node::aconst(value, (uintptr_t) table);
+
+   TR::ILOpCodes shiftOp = table->getDataType() == TR::Int32 ? TR::iushr : TR::lushr;
```
Comment (andrewcraik) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: might this look and read better as a setup of the whole thing for say Int64 and then a if Int32 overwrite them all and if something else assert? The repeated testing is a bit ugly and will be slower than necessary.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.hpp [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+      TR::TreeTop *insertionPoint,
+      TR::Node *value,
+      TR_AbstractHashTableProfilerInfo *table,
+      TR::Node *optionalTest = NULL,
+      TR::Node *fallbackValue = NULL,
+      bool extendBlocks = true,
+      bool trace = false);
+
+   private:
+   static const size_t DEFAULT_TABLE_BITS = 2;
```
Comment (andrewcraik) [inline runtime/compiler/trj9/optimizer/ValueJProfiling.hpp]: Can we have a comment on what this means here please?

File: runtime/compiler/trj9/runtime/J9ValueProfiler.hpp [hunk 27]
```diff
@@ -42,6 +42,8 @@ extern TR::Monitor *vpMonitor;
 
 #define ARRAY_MAX_NUM_VALUES (5)
 #define LINKEDLIST_MAX_NUM_VALUES (20)
+#define HASHTABLE_MAX_BITS  (8)
```
Comment (andrewcraik) [inline runtime/compiler/trj9/runtime/J9ValueProfiler.hpp]: comments on these are warranted

File: runtime/compiler/trj9/runtime/ValueProfiler.cpp [hunk 28]
```diff
@@ -188,13 +207,66 @@ TR_LinkedListProfilerInfo<TR_ByteInfo>::trace(TR::Compilation *comp)
...
+   }
+
 #if defined(TR_HOST_POWER)
 JIT_HELPER(_jitProfileValue);
 JIT_HELPER(_jitProfileLongValue);
 JIT_HELPER(_jitProfileBigDecimalValue);
 JIT_HELPER(_jitProfileStringValue);
 JIT_HELPER(_jitProfileAddress);
 JIT_HELPER(_jitProfileWarmCompilePICAddress);
+JIT_HELPER(_jProfileValue);
```
Comment (andrewcraik) [inline runtime/compiler/trj9/runtime/ValueProfiler.cpp]: should we call this jProfileIntValue? That way if we contribute it up to OMR where profiling of sub int values may be warranted the additions don't require refactoring?

**Comment by andrewcraik** [comment]
The overall design of this looks good - a few comments to address and a few things to clean up but otherwise looks good. We will need to run full Jenkins test once the ci issues are fixed because of the risk of things going wrong with the scale of this change.

File: runtime/compiler/trj9/codegen/J9TreeEvaluator.cpp (hunk 16 above)
Comment (ncough) [reply runtime/compiler/trj9/codegen/J9TreeEvaluator.cpp]: All API calls have been updated to use static casts.

File: runtime/compiler/trj9/compile/J9Compilation.hpp (hunk 17 above)
Comment (ncough) [reply runtime/compiler/trj9/compile/J9Compilation.hpp]: The intention was to treat them both as external value info in a single list, rather than adding one for each source of external info. However, I am not able to find where the hardware profiler hooks in to populate the original collection.

If hardware profiling were to be added again, it would have to add `TR_ExternalValueProfileInfo` on demand (as it would have with the original list) and potentially add special handling to `TR_ValueProfileInfoManager`, just as the IProfiler current does.

File: runtime/compiler/trj9/runtime/J9ValueProfiler.hpp (hunk 21 above)
Comment (ncough) [reply runtime/compiler/trj9/runtime/J9ValueProfiler.hpp]: Originally, these were assigned individual bits, however, that ultimately wasn't required. These have been corrected to use the default numbering.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp (hunk 24 above)
Comment (ncough) [reply runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: Yes, `getProperConversion` has a unsigned option I wasn't aware of. This has been corrected.

File: runtime/compiler/trj9/runtime/ValueProfiler.cpp (hunk 28 above)
Comment (ncough) [reply runtime/compiler/trj9/runtime/ValueProfiler.cpp]: These have been corrected to use the `jProfileXXBitValue` naming.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (ncough) [reply runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: Doxygen comments have been added to these methods.

File: runtime/compiler/trj9/optimizer/ValueJProfiling.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (ncough) [reply runtime/compiler/trj9/optimizer/ValueJProfiling.cpp]: This has been corrected to use `MAX_COLD_BLOCK_COUNT + 1`.

File: runtime/compiler/trj9/optimizer/BlockJProfiling.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (ncough) [reply runtime/compiler/trj9/optimizer/BlockJProfiling.cpp]: The functionality to abort the compile during a late swap to profiling has been added as a later commit to this PR.

File: runtime/compiler/trj9/compile/J9Compilation.hpp (hunk 17 above)
Comment (andrewcraik) [reply runtime/compiler/trj9/compile/J9Compilation.hpp]: @dsouzai you might be interested in this part of the change - any comments?

**Comment by andrewcraik** [comment]
@mpirvu FYI - addition of the exception to restart compilation with a switch to profiling will be of interest to you - a review on that aspect/commit from you would be good.

File: runtime/compiler/trj9/compile/J9Compilation.hpp (hunk 17 above)
Comment (dsouzai) [reply runtime/compiler/trj9/compile/J9Compilation.hpp]: I don't think anything was using `getMethodHWVPInfos`. Even when we were experimenting with Value Profiling with HWProfiling, we used these APIs 

https://github.com/eclipse/openj9/blob/2a24b14d040312a0ec54075a171070558a71e3be/runtime/compiler/trj9/compile/J9Compilation.hpp#L217-L231

**Comment by ncough** [comment]
This change depends on: https://github.com/eclipse/omr/pull/2121

**Comment by ncough** [comment]
After a discussion with @mpirvu, the component of this change, capable of identifying and correcting hash tables that have missed the most frequent values, will be moved to a separate pull request.

**Comment by andrewcraik** [comment]
@ncough could you comment on what you have updated once you are ready for the review to continue since I would like to know where to focus further review and if anyone else needs to asked to comment on the changes so we can keep moving on this review since this is a fairly big set of changes. Thank you very much for your proposed contribution!

**Comment by ncough** [comment]
The change set has undergone a few modifications since the last round of reviews. The most significant changes are:

- Move to JProfilingBlock/JProfilingValue naming.
- Modifications to the JProfilingValue trees to correct behaviour for loads and zero extension of bytes and shorts, which were problematic on Z.
- Use of system linkage on Z to call the JProfilingValue helper (@r30shah).
- Use of `VM_AtomicSupport` to implement `tryLock`, `lock` and `unlock` on accesses to the hash tables.
- Minor changes to hash table data structures to enable compilation on AIX and ZOS, such as issues equating `uintptr_t` to one of `uint32_t` or `uint64_t`.
- More aggressive cleanup of persistent profile information, freed once their reference counts hit zero with synchronization using `VM_AtomicSupport` methods.
- Tracing for compilations restarted due to `switchToProfiling`.
- Disabled sampling and counted recompilation triggers, instead relying on JProfiling infrastructure to trigger the recompilation.

A review from @dsouzai on the two control change would be greatly appreciated.

File: runtime/compiler/trj9/runtime/J9Profiler.cpp [hunk 29]
```diff
@@ -2448,3 +2507,95 @@ void TR_PersistentProfileInfo::dumpInfo(TR::FILE *logFile)
...
+
+   TR::Recompilation *recompilation = comp->getRecompilationInfo();
+   if (recompilation)
+      {
+      TR_PersistentMethodInfo *methodInfo = recompilation->getMethodInfo();
+      _current = compare(methodInfo);
+      if (_current && _current == TR_PersistentProfileInfo::getCurrent(comp))
+         {
+         TR_PersistentProfileInfo::decRefCount(_current); 
+         _current = NULL;
```
Comment (dsouzai) [inline runtime/compiler/trj9/runtime/J9Profiler.cpp]: Does this mean that the only available profiling information is from the current compile, and, since it hasn't executed anything (because it's still being compiled), really there is no profiling information?

File: runtime/compiler/trj9/control/RecompilationInfo.hpp [hunk 30]
```diff
@@ -372,6 +390,8 @@ class TR_PersistentJittedBodyInfo
 
    bool       isLongRunningInterpreted() const { return _longRunningInterpreted; }
 
+   void setProfileInfo(TR_PersistentProfileInfo * ppi) { _profileInfo = ppi; }
+   TR_PersistentProfileInfo *getProfileInfo() { return _profileInfo; }
```
Comment (dsouzai) [inline runtime/compiler/trj9/control/RecompilationInfo.hpp]: Isn't it possible that if someone acquires a reference to the bodyInfo indirectly (ie via the J9JITExceptionTable) and then decides to dereference the profileInfo, because `getProfileInfo` doesn't go through the process of acquiring the lock (and thus doing a ref count inc), they could try and access freed memory? If so, it is worth adding a comment here to state that `getProfileInfo` should only be called via the `TR_PersistentMethodInfo` or `J9::Recompilation` objects, as they will do the proper work of locking and incrementing the ref count.

File: runtime/compiler/trj9/runtime/J9Profiler.hpp [hunk 31]
```diff
@@ -91,7 +104,8 @@ class TR_PersistentProfileInfo
         _catchBlockProfileInfo(NULL),
         _blockFrequencyInfo(NULL),
         _valueProfileInfo(NULL),
-        _callSiteInfo(NULL)
+        _callSiteInfo(NULL),
+        _refCount(1)
```
Comment (dsouzai) [inline runtime/compiler/trj9/runtime/J9Profiler.hpp]: You should explicitly document what can increment the ref counts. For example, in the constructor of `TR_PersistentProfileInfo`, the refcount is set to 1 because the `TR_PersistentBodyInfo` creates a `TR_PersistentProfileInfo` and hence has a reference to it. The other three scenarios I can think of are:

* _recentProfileInfo in `TR_PersistentMethodInfo`
* _bestProfileInfo in `TR_PersistentMethodInfo`
* Some compilation thread via `TR_AccessedProfileInfo`

File: runtime/compiler/trj9/control/RecompilationInfo.hpp (hunk 30 above)
Comment (ncough) [reply runtime/compiler/trj9/control/RecompilationInfo.hpp]: Yes, you are right. These methods rely on the profile information persisting at least until `jitReleaseCodeCollectMetaData` cleans up the meta data for a body, setting this pointer to `NULL` in the process. Uses of these calls can all guarantee that they are safe, as they will remain within this duration, such as when the body is actively being compiled. I'll add a comment documenting this behaviour.

File: runtime/compiler/trj9/runtime/J9Profiler.cpp (hunk 29 above)
Comment (ncough) [reply runtime/compiler/trj9/runtime/J9Profiler.cpp]: Yes, there is no point inspecting this information and it may mislead uses of profiling data.

File: runtime/compiler/trj9/runtime/J9Profiler.hpp (hunk 31 above)
Comment (ncough) [reply runtime/compiler/trj9/runtime/J9Profiler.hpp]: I have extended the comment above to include the additional details.

**Comment by andrewcraik** [comment]
This depends on https://github.com/eclipse/omr/pull/2121

**Comment by andrewcraik** [comment]
Jenkins test sanity depends eclipse/omr#2121

**Comment by ncough** [comment]
Both sanity tests passed, however, there was a configuration issue with profiling compilations. If `enableJProfiling` was set and the compilation performed a late swap to profiling, it would add the prior implementation of block profiling in addition to JProfilingBlock. As these two make different assumptions about the same data structure, it would result in a crash.

**Comment by andrewcraik** [comment]
Jenkins test sanity depends eclipse/omr#2121
